Inner public template class leads to NullPointerException

I've minimized failure to
```
struct A
  {
  template<class T>class B;
  };
```

It works with class and private declaration
```
class A
  {
  template<class T>class B;
  };
```

but fails with public:
```
class A
  {
  public:
    template<class T>class B;
  };
```

org.sonar.squidbridge.api.AnalysisException: Unable to analyze file: ...
        at org.sonar.squidbridge.AstScanner.scanFiles(AstScanner.java:131)
        at org.sonar.plugins.cxx.squid.CxxSquidSensor.execute(CxxSquidSensor.java:141)
        at org.sonar.batch.sensor.SensorWrapper.analyse(SensorWrapper.java:57)
        at org.sonar.batch.phases.SensorsExecutor.executeSensor(SensorsExecutor.java:58)
        at org.sonar.batch.phases.SensorsExecutor.execute(SensorsExecutor.java:50)
        at org.sonar.batch.phases.AbstractPhaseExecutor.execute(AbstractPhaseExecutor.java:83)
        at org.sonar.batch.scan.ModuleScanContainer.doAfterStart(ModuleScanContainer.java:192)
        at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
        at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
        at org.sonar.batch.scan.ProjectScanContainer.scan(ProjectScanContainer.java:241)
        at org.sonar.batch.scan.ProjectScanContainer.scanRecursively(ProjectScanContainer.java:236)
        at org.sonar.batch.scan.ProjectScanContainer.scanRecursively(ProjectScanContainer.java:234)
        at org.sonar.batch.scan.ProjectScanContainer.doAfterStart(ProjectScanContainer.java:226)
        at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
        at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
        at org.sonar.batch.task.ScanTask.execute(ScanTask.java:47)
        at org.sonar.batch.task.TaskContainer.doAfterStart(TaskContainer.java:86)
        at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
        at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
        at org.sonar.batch.bootstrap.GlobalContainer.executeTask(GlobalContainer.java:106)
        at org.sonar.batch.bootstrapper.Batch.executeTask(Batch.java:119)
        at org.sonarsource.scanner.api.internal.batch.BatchIsolatedLauncher.execute(BatchIsolatedLauncher.java:62)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at org.sonarsource.scanner.api.internal.IsolatedLauncherProxy.invoke(IsolatedLauncherProxy.java:60)
        at com.sun.proxy.$Proxy0.execute(Unknown Source)
        at org.sonarsource.scanner.api.EmbeddedScanner.doExecute(EmbeddedScanner.java:233)
        at org.sonarsource.scanner.api.EmbeddedScanner.runAnalysis(EmbeddedScanner.java:151)
        at org.sonarsource.scanner.cli.Main.runAnalysis(Main.java:110)
        at org.sonarsource.scanner.cli.Main.execute(Main.java:74)
        at org.sonarsource.scanner.cli.Main.main(Main.java:61)
Caused by: java.lang.NullPointerException
        at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitTemplateDeclaration(AbstractCxxPublicApiVisitor.java:488)
        at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitMemberDeclaration(AbstractCxxPublicApiVisitor.java:432)
        at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitNode(AbstractCxxPublicApiVisitor.java:153)
        at com.sonar.sslr.impl.ast.AstWalker.visitNode(AstWalker.java:114)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:85)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.walkAndVisit(AstWalker.java:69)
        at org.sonar.squidbridge.AstScanner.scanFiles(AstScanner.java:110)
... 32 more

--- a/cxx-squid/src/main/java/org/sonar/cxx/visitors/AbstractCxxPublicApiVisitor.java
+++ b/cxx-squid/src/main/java/org/sonar/cxx/visitors/AbstractCxxPublicApiVisitor.java
@@ -204,12 +204,12 @@ public abstract class AbstractCxxPublicApiVisitor<GRAMMAR extends Grammar>
     if (memberDeclaration != null) {
       return;
     }
-    
+
     // ignore classSpefifier typedefs (classSpefifier at classSpefifier)
     if (isTypedef(declaratorList)) {
       return;
     }
-    
+
     // ignore friend declarations
     if (isFriendDeclarationList(declaratorList)) {
       return;
@@ -254,7 +254,7 @@ public abstract class AbstractCxxPublicApiVisitor<GRAMMAR extends Grammar>
     }
     return false;
   }
-  
+
   private boolean isFriendDeclarationList(AstNode declaratorList) {
     AstNode simpleDeclNode = declaratorList
       .getFirstAncestor(CxxGrammarImpl.simpleDeclaration);
@@ -353,7 +353,7 @@ public abstract class AbstractCxxPublicApiVisitor<GRAMMAR extends Grammar>
     }
     return null;
   }
-  
+
   private void visitClassSpecifier(AstNode classSpecifier) {
 
     // check if this is a template specification to adjust documentation node
@@ -365,7 +365,7 @@ public abstract class AbstractCxxPublicApiVisitor<GRAMMAR extends Grammar>
         docNode = classSpecifier;
       }
     }
-    
+
     // narrow the identifier search scope to classHead
     AstNode classHead = classSpecifier
       .getFirstDescendant(CxxGrammarImpl.classHead);
@@ -477,15 +477,18 @@ public abstract class AbstractCxxPublicApiVisitor<GRAMMAR extends Grammar>
 
     return false;
   }
-    
+
   private void visitTemplateDeclaration(AstNode templateDeclaration) {
 
     if (isFriendMemberDeclaration(templateDeclaration.getParent())) {
       return;
     }
 
-    String id = templateDeclaration.getFirstDescendant(CxxGrammarImpl.className)
-      .getTokenValue();
+    AstNode className = templateDeclaration.getFirstDescendant(CxxGrammarImpl.className);
+    if (className == null) {
+      return;
+    }
+    String id = className.getTokenValue();
 
     // handle cascaded template declarations
     AstNode node = templateDeclaration;
@@ -498,10 +501,10 @@ public abstract class AbstractCxxPublicApiVisitor<GRAMMAR extends Grammar>
       }
       node = node.getFirstAncestor(CxxGrammarImpl.templateDeclaration);
     } while (node != null);
-    
+
     visitPublicApi(templateDeclaration, id, comments);
   }
-  
+
   private void visitFunctionDefinition(AstNode functionDef) {
     if (isPublicApiMember(functionDef)) {
       // filter out deleted and defaulted methods
@@ -570,7 +573,7 @@ public abstract class AbstractCxxPublicApiVisitor<GRAMMAR extends Grammar>
     AstNode container = node.getFirstAncestor(
       CxxGrammarImpl.templateDeclaration,
       CxxGrammarImpl.classSpecifier);
-    
+
     AstNode docNode = node;
     List<Token> comments;
 
@@ -629,8 +632,8 @@ public abstract class AbstractCxxPublicApiVisitor<GRAMMAR extends Grammar>
       CxxGrammarImpl.classSpecifier);
 
     // An alias declaration inside a function is not part of the public API
-    if (parent != null && parent.getType() == CxxGrammarImpl.functionDefinition){
-        return;
+    if (parent != null && parent.getType() == CxxGrammarImpl.functionDefinition) {
+      return;
     }
 
     if (isPublicApiMember(aliasDeclNode)) {
@@ -687,7 +690,7 @@ public abstract class AbstractCxxPublicApiVisitor<GRAMMAR extends Grammar>
     if (docNode == null) {
       docNode = enumSpecifierNode;
     }
-      
+
     visitPublicApi(
       enumSpecifierNode, enumId,
       getBlockDocumentation(docNode));
@@ -778,7 +781,7 @@ public abstract class AbstractCxxPublicApiVisitor<GRAMMAR extends Grammar>
 
   private static boolean isPublicApiMember(AstNode node) {
     AstNode access = node;
-    
+
     // retrieve the accessSpecifier
     do {
       access = access.getPreviousAstNode();
@@ -816,9 +819,9 @@ public abstract class AbstractCxxPublicApiVisitor<GRAMMAR extends Grammar>
         } else {
           LOG.error("isPublicApiMember: failed to get enclosing classSpecifier for node at {}", node.getTokenLine());
           return false;
-        }    
+        }
       }
-    
+
       // method or function outside of class
       return true;
     }

--- a/cxx-squid/src/test/java/org/sonar/cxx/visitors/CxxPublicApiVisitorTest.java
+++ b/cxx-squid/src/test/java/org/sonar/cxx/visitors/CxxPublicApiVisitorTest.java
@@ -137,7 +137,7 @@ public class CxxPublicApiVisitorTest {
 
   @Test
   public void template() throws IOException {
-    testFile("src/test/resources/metrics/template.h", 13, 4, false);
+    testFile("src/test/resources/metrics/template.h", 14, 4, false);
   }
 
   @Test

--- a/cxx-squid/src/test/resources/metrics/template.h
+++ b/cxx-squid/src/test/resources/metrics/template.h
@@ -64,3 +64,10 @@ template <class T>
 void TestClass2::functionTest2()
 {
 }
+
+/**
+ * @brief issue #1067
+ */
+struct A {
+   template<class T> class B;
+};

